sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/json/JsonSuite.scala

```diff
@@ -2065,29 +2065,238 @@ class JsonSuite extends QueryTest with SharedSQLContext with TestJsonData {
+        .save(path.getCanonicalPath)
+
+      checkCharset(
+        expectedCharset = "UTF-8",
+        pathToJsonFiles = path.getCanonicalPath,
+        expectedContent = """{"_1":"Dog","_2":42}"""
+      )
+    }
+  }
+
+  test("SPARK-23723: wrong output charset") {
+    val charset = "UTF-128"
+    val exception = intercept[java.io.UnsupportedEncodingException] {
+      withTempPath { path =>
+        val df = spark.createDataset(Seq((0)))
+        df.write
+          .options(Map("charset" -> charset, "lineSep" -> "\n"))
+          .format("json").mode("overwrite")
+          .save(path.getCanonicalPath)
+      }
+    }
+
+    assert(exception.getMessage == charset)
+  }
+
+  test("SPARK-23723: read written json in UTF-16") {
+    val charset = "UTF-16"
+    case class Rec(f1: String, f2: Int)
+    withTempPath { path =>
+      val ds = spark.createDataset(Seq(
+        ("a", 1), ("b", 2), ("c", 3))
+      ).repartition(2)
+      ds.write
+        .options(Map("charset" -> charset, "lineSep" -> "\n"))
+        .format("json").mode("overwrite")
+        .save(path.getCanonicalPath)
+      val savedDf = spark
+        .read
+        .schema(ds.schema)
+        // This option will be replaced by .option("lineSep", "x00 0a")
+        // as soon as lineSep allows to specify sequence of bytes in hexadecimal format.
+        .option("mode", "DROPMALFORMED")
+        .json(path.getCanonicalPath)
+
+      checkAnswer(savedDf.toDF(), ds.toDF())
+    }
+  }
+
+  def checkReadJson(
+    lineSep: String,
+    charsetOption: String,
+    charset: String,
+    inferSchema: Boolean,
+    runId: Int
+  ): Unit = {
+    test(s"SPARK-23724: checks reading json in ${charset} #${runId}") {
+      val delimInBytes = {
+        if (lineSep.startsWith("x")) {
+          lineSep.replaceAll("[^0-9A-Fa-f]", "")
+            .sliding(2, 2).toArray.map(Integer.parseInt(_, 16).toByte)
+        } else {
+          lineSep.getBytes(charset)
         }
       }
+      case class Rec(f1: String, f2: Int) {
+        def json = s"""{"f1":"${f1}", "f2":$f2}"""
+        def bytes = json.getBytes(charset)
+        def row = Row(f1, f2)
+      }
+      val schema = new StructType().add("f1", StringType).add("f2", IntegerType)
+      withTempPath { path =>
+        val records = List(Rec("a", 1), Rec("b", 2))
+        val data = records.map(_.bytes).reduce((a1, a2) => a1 ++ delimInBytes ++ a2)
+        val os = new FileOutputStream(path)
+        os.write(data)
+        os.close()
+        val reader = if (inferSchema) {
+          spark.read
+        } else {
+          spark.read.schema(schema)
+        }
+        val savedDf = reader
+          .option(charsetOption, charset)
+          .option("lineSep", lineSep)
+          .json(path.getCanonicalPath)
+        checkAnswer(savedDf, records.map(_.row))
+      }
+    }
+  }
 
+  // scalastyle:off nonascii
+  List(
+    ("|", "encoding", "UTF-8", false),
+    ("^", "charset", "UTF-16BE", true),
```

Comment:
Shall we have one test for `charset` and use `encoding` everywhere?

Reply:
done